lib/charms/glauth_k8s/v0/ldap.py

```diff
@@ -0,0 +1,325 @@
+        self.app = charm.app
+        self.unit = charm.unit
+        self._relation_name = relation_name
+
+        self.framework.observe(
+            self.charm.on[self._relation_name].relation_changed,
+            self._on_relation_changed,
+        )
+
+    @leader_unit
+    def _on_relation_changed(self, event: RelationChangedEvent) -> None:
+        """Handle the event emitted when the consumer charm provides the
+        necessary data."""
+
+        self.on.ldap_requested.emit(event.relation)
+
+    def update_relation_app_data(
+        self, relation_id: int, data: LdapProviderData
+    ) -> None:
+        """An API for the provider charm to provide the LDAP related
+        information."""
+
+        relation = self.charm.model.get_relation(
+            self._relation_name, relation_id
+        )
+        _update_relation_app_databag(self.charm, relation, asdict(data))
+
+
+class LdapConsumer(Object):
+    """An LDAP consumer to consume data delivered by an LDAP provider charm."""
+
+    on = LdapConsumerEvents()
+
+    def __init__(
+        self,
+        charm: CharmBase,
+        relation_name: str = DEFAULT_RELATION_NAME,
+    ) -> None:
+        super().__init__(charm, relation_name)
+
+        self.charm = charm
+        self.app = charm.app
+        self.unit = charm.unit
+        self._relation_name = relation_name
+
+        self.framework.observe(
+            self.charm.on[self._relation_name].relation_created,
+            self._on_ldap_relation_created,
+        )
+        self.framework.observe(
+            self.charm.on[self._relation_name].relation_changed,
+            self._on_ldap_relation_changed,
+        )
+        self.framework.observe(
+            self.charm.on[self._relation_name].relation_broken,
+            self._on_ldap_relation_broken,
+        )
+
+    def _on_ldap_relation_created(self, event: RelationCreatedEvent) -> None:
+        """Handle the event emitted when an LDAP integration is created."""
+
+        app = self.app.name
+        model = self.model.name
+        _update_relation_app_databag(
+            self.charm, event.relation, {"app": app, "model": model}
+        )
+
+    def _on_ldap_relation_changed(self, event: RelationChangedEvent) -> None:
+        """Handle the event emitted when the LDAP related information is
+        ready."""
+
+        provider_app = event.relation.app
+
+        if not event.relation.data.get(provider_app):
+            return
+
+        self.on.ldap_ready.emit(event.relation)
+
+    def _on_ldap_relation_broken(self, event: RelationBrokenEvent) -> None:
+        """Handle the event emitted when the LDAP integration is broken."""
+
+        self.on.ldap_unavailable.emit(event.relation)
+
+    @staticmethod
+    def consume_ldap_relation_data(
+        relation: Relation,
+    ) -> Optional[LdapProviderData]:
+        """An API for the consumer charm to consume the LDAP related
+        information in the application databag."""
+
```

Comment:
I don't think this should be a static method. The class object already has the relation name, we can get the relation from there
```suggestion
    def consume_ldap_relation_data(self, relation_id: Optional[int]) -> Optional[LdapProviderData]:
        """An API for the consumer charm to consume the LDAP related
        information in the application databag."""

        if len(self.model.relations) == 0:
            return

        relation = self.model.get_relation(self._relation_name, relation_id=relation_id)
        if not relation or not relation.app:
            return
```